system/include/emscripten/emscripten.h

```diff
@@ -168,6 +168,11 @@ typedef void (*em_scan_func)(void*, void*);
 void emscripten_scan_registers(em_scan_func func);
 void emscripten_scan_stack(em_scan_func func);
 
+// Asyncronous version of dlopen.  Since WebAssembly module loading in general
+// is asyncronous the normal dlopen function can't be used in all situations.
+typedef void (*em_dlopen_callback)(void* handle, void* user_data);
```

Comment:
I guess one question i have is should we support returning non-zero here to indicate synchronous failure?

Reply:
I don't follow - what else can we do, on such a failure, but return an error?

Reply:
Right now we are returning void, and calling using the `onerror` handler as the one and only way to signal and error.
For example, if we get given invalid arguments with the current API the `onerror` handler is the only way we get notified about that failure.

I'm suggesting that maybe we could also declare that a non-zero return from emscripten_dlopen can be used to indicate an synchronous failure.  i.e. that request could not even be iniated for some reason.    I guess we should follow whatever the existing pattern is for fetch/wget.

Reply:
I see, thanks. I think `void` is fine, as in wget.